### PR TITLE
Guard on default device

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99551

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -452,6 +452,15 @@ def guard_source(self):
         return GuardSource.GLOBAL
 
 
+@dataclasses.dataclass(frozen=True)
+class DefaultDeviceSource(Source):
+    def name(self):
+        return ""
```
Comment (voznesenskym) [inline torch/_dynamo/source.py:458]: Just don't define it

Comment (voznesenskym) [reply torch/_dynamo/source.py:458]: Actually, we can't notimplemented this, because of provided_guarded_object.

Okay.

**Review by voznesenskym — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

**Comment by ezyang** [comment]
[Perf results](hud.pytorch.org/benchmark/compilers?startTime=Mon, 17 Apr 2023 12:54:32 GMT&stopTime=Thu, 20 Apr 2023 12:54:32 GMT&suite=torchbench&mode=training&dtype=amp&lBranch=gh/ezyang/2018/head&lCommit=cf8dfdc5f57f8d19efa9d1d784fb4bb25430d3c6&rBranch=main&rCommit=0ae9d15543fbc7636f37d4be6f01a1b9e84159dc)

**Pull request closed without merging**